optaplanner-core/src/main/java/org/optaplanner/core/impl/constructionheuristic/DefaultConstructionHeuristicPhaseFactory.java

```diff
@@ -75,23 +83,10 @@ public DefaultConstructionHeuristicPhaseFactory(ConstructionHeuristicPhaseConfig
                 .withValueSorterManner(valueSorterManner)
                 .build();
         Termination<Solution_> phaseTermination = buildPhaseTermination(phaseConfigPolicy, solverTermination);
-        EntityPlacerConfig entityPlacerConfig_;
-        if (phaseConfig.getEntityPlacerConfig() == null) {
-            entityPlacerConfig_ = buildUnfoldedEntityPlacerConfig(phaseConfigPolicy, constructionHeuristicType_);
-        } else {
-            entityPlacerConfig_ = phaseConfig.getEntityPlacerConfig();
-            if (phaseConfig.getConstructionHeuristicType() != null) {
-                throw new IllegalArgumentException(
-                        "The constructionHeuristicType (" + phaseConfig.getConstructionHeuristicType()
-                                + ") must not be configured if the entityPlacerConfig (" + entityPlacerConfig_
-                                + ") is explicitly configured.");
-            }
-            if (phaseConfig.getMoveSelectorConfigList() != null) {
-                throw new IllegalArgumentException("The moveSelectorConfigList (" + phaseConfig.getMoveSelectorConfigList()
-                        + ") cannot be configured if the entityPlacerConfig (" + entityPlacerConfig_
-                        + ") is explicitly configured.");
-            }
-        }
+        EntityPlacerConfig entityPlacerConfig_ = getValidEntityPlacerConfig()
+                .orElseGet(() -> getValidListVariableDescriptor(solverConfigPolicy)
+                        .map(variableDescriptor -> buildListVariablePlacerConfig(phaseConfigPolicy, variableDescriptor))
+                        .orElseGet(() -> buildUnfoldedEntityPlacerConfig(phaseConfigPolicy, constructionHeuristicType_)));
```

Comment:
I am not sure this nesting of `orElseGet()` helps with readability. Consider a small change:

    EntityPlacerConfig entityPlacerConfig_ = getValidEntityPlacerConfig()
                .orElseGet(this::buildEntityPlacerConfigIfNoneExists);

    ....

    private EntityPlacerConfig buildEntityPlacerConfigIfNoneExists() {
        return getValidListVariableDescriptor(solverConfigPolicy)
                        .map(variableDescriptor -> buildListVariablePlacerConfig(phaseConfigPolicy, variableDescriptor))
                        .orElseGet(() -> buildUnfoldedEntityPlacerConfig(phaseConfigPolicy, constructionHeuristicType_));
    }

Not only is this easier to read I think, it also allows you to unit-test the nested method, if necessary.
In general, lambdas are awesome if they're short and concise. For multi-line lambdas, method references often work better.

Reply:
Yes, that would definitely be easier to read. Unfortunately, `solverConfigPolicy`, `phaseConfigPolicy` and `constructionHeuristicType_` are all local variables (not fields). I'd need to make them parameters of `buildEntityPlacerConfigIfNoneExists()` and that thwarts the method reference approach.

But I agree the nested lambdas are horrible to read. I'll look if there's another way to improve this part.